### PR TITLE
added custom fields to user model

### PR DESCRIPTION
**Pull request by ElijahAhianyo**
<a href="https://gitpod.io/#https://github.com/Xcov19/covidX/pull/69"></a>

---
This change is [Reviewable](https://reviewable.io/reviews/xcov19/covidx/69)

<a href="https://gitpod.io/#https://github.com/Xcov19/covidX/pull/69"></a>

### CODE REVIEW

File: apps/auth_zero/migrations/0004_auto_20201218_1031.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    operations = [
+        migrations.AddField(
+            model_name="user",
+            name="last_otp",
+            field=models.CharField(max_length=20, null=True),
+        ),
+        migrations.AddField(
+            model_name="user",
+            name="mobile",
+            field=models.CharField(
```
Comment (codecakes) [inline apps/auth_zero/migrations/0004_auto_20201218_1031.py:22]: install  https://pypi.org/project/Babel/ and use [phonenumbers](https://github.com/stefanfoulis/django-phonenumber-field) which is already installed in the project.
add PHONENUMBER_DEFAULT_REGION to `.env`

File: apps/auth_zero/models.py [hunk 2]
```diff
@@ -26,6 +27,15 @@ class User(AbstractUser):
     """Custom Auth User model."""
 
     default_user_type = UserTypes.PATIENT
+    phone_regex = RegexValidator(
```
Comment (codecakes) [inline apps/auth_zero/models.py:30]: this becomes obsolete when we use django-phone numbers.

File: apps/auth_zero/models.py [hunk 3]
```diff
@@ -26,6 +27,15 @@ class User(AbstractUser):
...
+        regex=r"^\+?1?\d{9,15}$",
+        message="""Phone number must be entered in the format:
+         '+999999999'. Up to 15 digits allowed.""",
+    )
+    mobile = models.CharField(validators=[phone_regex], max_length=17)
```
Comment (codecakes) [inline apps/auth_zero/models.py:35]: use **PhoneNumberField()**

File: apps/auth_zero/models.py [hunk 4]
```diff
@@ -26,6 +27,15 @@ class User(AbstractUser):
...
+    mobile_is_verified = models.BooleanField(default=False)
```
Comment (codecakes) [inline apps/auth_zero/models.py:36]: do we really need this if we have `otp_verified`?

**Review by codecakes — CHANGES_REQUESTED** [review]
also write some unit tests creating and committing models with some edge cases for this model.
cases are when email is verified but mobile num is not and vice versa.
unless both of them are not verified user is not verified.
for that purpose, you could do something like:
```python3

class User:
  ..
  ..
  @property
  def is_verified(self):
    return self.otp_verified and self.email_verified
```

check again after running shell_plus or from DB model schema, i think its self.auth_user.email_verified ;

**Comment by codecakes** [comment]
@ElijahAhianyo if working on JIRA tasks, use [smart commits ](https://confluence.atlassian.com/fisheye/using-smart-commits-960155400.html)

**Comment by codecakes** [comment]
@ElijahAhianyo first un apply this particular migration and then remigrate after making changes

**Comment by milan-tom** [comment]
> also write some unit tests creating and committing models with some edge cases for this model.
> cases are when email is verified but mobile num is not and vice versa.
> unless both of them are not verified user is not verified.
> for that purpose, you could do something like:
> ```python3
> 
> class User:
>   ..
>   ..
>   @property
>   def is_verified(self):
>     return self.otp_verified and self.email_verified
> ```
> 
> check again after running shell_plus or from DB model schema, i think its self.auth_user.email_verified ; 

@ElijahAhianyo I can do this section. You can focus on the rest of the requested changes.

File: apps/auth_zero/models.py (hunk 4 above)
Comment (ElijahAhianyo) [reply apps/auth_zero/models.py:36]: I kind of added that because it was specified in the list of custom fields(on jira) to be added. If it's not useful, I can as well take that out.

Comment (codecakes) [reply apps/auth_zero/models.py:36]: yes remove it.

File: covidX/settings.py [hunk 5]
```diff
@@ -152,6 +153,7 @@
         }
     }
 else:
+    print('yay')
```
Comment (codecakes) [inline covidX/settings.py:156]: `print('yay')` @ElijahAhianyo ??

**Review by codecakes — CHANGES_REQUESTED** [review]
would great if @milan-tom could merge his pr https://github.com/ElijahAhianyo/covidX/pull/1 to your branch @ElijahAhianyo

**Comment by codecakes** [comment]
> ## DeepCode's analysis on [#e8374e](https://www.deepcode.ai/app/gh/Xcov19/covidX/1a1ed3150c827794748e0734ef6cec8b8d4ce8ab/Xcov19/covidX/e8374e5299a0abc9636393f8968be082ecde90ab/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=2&) found:
> 
>     * ℹ️ **2** minor issues. 👇
> 
> 
> ## Top issues
> Description 	Example fixes
> Do not hardcode credentials in code. Found hardcoded credential used in username. Occurrences:
> 
>     * [tests.py:11](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L11)
> 
>     * [tests.py:21](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L21)
> 
>     * [tests.py:32](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L32)
> 
> 
> 	[🔧 Example fixes](https://www.deepcode.ai/app/gh/Xcov19/covidX/1a1ed3150c827794748e0734ef6cec8b8d4ce8ab/Xcov19/covidX/e8374e5299a0abc9636393f8968be082ecde90ab/pr/_/%2Fapps%2Fauth_zero%2Ftests.py/python%2Fdc%2FNoHardcodedCredentials%2Ftest/code/?utm_source=gh_review&)
> Do not hardcode passwords in code. Found hardcoded password used in password. Occurrences:
> 
>     * [tests.py:13](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L13)
> 
>     * [tests.py:23](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L23)
> 
>     * [tests.py:34](https://github.com/Xcov19/covidX/blob/e8374e5299a0abc9636393f8968be082ecde90ab/apps/auth_zero/tests.py#L34)
> 
> 
> 	[🔧 Example fixes](https://www.deepcode.ai/app/gh/Xcov19/covidX/1a1ed3150c827794748e0734ef6cec8b8d4ce8ab/Xcov19/covidX/e8374e5299a0abc9636393f8968be082ecde90ab/pr/_/%2Fapps%2Fauth_zero%2Ftests.py/python%2Fdc%2FNoHardcodedPasswords%2Ftest/code/?utm_source=gh_review&)
> #### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/Xcov19/covidX/1a1ed3150c827794748e0734ef6cec8b8d4ce8ab/Xcov19/covidX/e8374e5299a0abc9636393f8968be082ecde90ab/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=2&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=Xcov19)

@milan-tom do the following to fix the ☝️ issue and bring more structure and clarity to overall testing :
- create config/__init__.py folder and move all credentials you are creating in a `test_config.ini` file
- Define sections as follows which will help you organize test data:

```ini

[DEFAULT]
USERNAME = 
PASSWORD = 
EMAIL = 
DB_USERNAME = 
DB_PASSWORD = 

[verified]
User =
Password

[unverified]
similarly here = 
```
- you can use `configparser` built-in lib to load from here like:
```python3

import six
import configparser

confg = configparser.ConfigParser()
confg.read(your ini file)

user = six.text_type(confg['unverified']['USER']) and so on..
```

File: apps/auth_zero/tests.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+from django.contrib.auth import get_user_model
+
+
+class UserTestCase(TestCase):
+    def setUp(self):
+        self.user_model = get_user_model()
+
+    def test_fully_unverified_user(self):
+        fully_unverified_user = self.user_model.objects.create_user(
+            username="fullyUnverifiedUser",
```
Comment (codecakes) [inline apps/auth_zero/tests.py:11]: @milan-tom see https://github.com/Xcov19/covidX/pull/69#issuecomment-753284121

File: apps/auth_zero/tests.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+from django.test import TestCase
+from django.contrib.auth import get_user_model
+
+
+class UserTestCase(TestCase):
+    def setUp(self):
+        self.user_model = get_user_model()
+
```
Comment (codecakes) [inline apps/auth_zero/tests.py:8]: load all credentials and values you are using during setup here; follow DRY

File: apps/auth_zero/tests.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        self.assertFalse(fully_unverified_user.is_verified)
+
+    def test_partially_verfied_users(self):
+        only_otp_verified_user = self.user_model.objects.create_user(
+            username="onlyOTPVerifiedUser",
+            email="onlyOTPVerifiedUser@mail.com",
+            password="onlyOTPVerifiedUserPassword"
+        )
+        only_otp_verified_user.save()
+        only_otp_verified_user.otp_verified = True
```
Comment (codecakes) [inline apps/auth_zero/tests.py:26]: Could `otp_verified = True` this be not included when you're creating the entry using create_user?

File: apps/auth_zero/tests.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        self.assertFalse(only_email_verified_user.is_verified)
+
+    def test_fully_verified_user(self):
+        fully_verified_user = self.user_model.objects.create_user(
+            username="fullyVerifiedUser",
+            email="fullyVerifiedUser@mail.com",
+            password="fullyVerifiedUserPassword"
+        )
+        fully_verified_user.save()
+        fully_verified_user.otp_verified = fully_verified_user.email_verified = True
```
Comment (codecakes) [inline apps/auth_zero/tests.py:49]: same as above. if we can remove them to create_user method?

**Review by codecakes — CHANGES_REQUESTED** [review]
@milan-tom left few comments. I've requested some changes. this will bring more structure as we bring test coverage to our codebase. i'll review again on those changes.

**Comment by milan-tom** [comment]
> @milan-tom do the following to fix the ☝️ issue and bring more structure and clarity to overall testing :
> 
> * create config/**init**.py folder and move all credentials you are creating in a `test_config.ini` file
> * Define sections as follows which will help you organize test data:
> 
> ```ini
> [DEFAULT]
> USERNAME = 
> PASSWORD = 
> EMAIL = 
> DB_USERNAME = 
> DB_PASSWORD = 
> 
> [verified]
> User =
> Password
> 
> [unverified]
> similarly here = 
> ```
> 
> * you can use `configparser` built-in lib to load from here like:
> 
> ```python
> import six
> import configparser
> 
> confg = configparser.ConfigParser()
> confg.read(your ini file)
> 
> user = six.text_type(confg['unverified']['USER']) and so on..
> ```

What is the purpose of `six.text_type` here?

**Comment by codecakes** [comment]
> > @milan-tom do the following to fix the ☝️ issue and bring more structure and clarity to overall testing :
> > 
> > * create config/**init**.py folder and move all credentials you are creating in a `test_config.ini` file
> > * Define sections as follows which will help you organize test data:
> > 
> > ```ini
> > [DEFAULT]
> > USERNAME = 
> > PASSWORD = 
> > EMAIL = 
> > DB_USERNAME = 
> > DB_PASSWORD = 
> > 
> > [verified]
> > User =
> > Password
> > 
> > [unverified]
> > similarly here = 
> > ```
> > 
> > 
> > 
> > * you can use `configparser` built-in lib to load from here like:
> > 
> > ```python
> > import six
> > import configparser
> > 
> > confg = configparser.ConfigParser()
> > confg.read(your ini file)
> > 
> > user = six.text_type(confg['unverified']['USER']) and so on..
> > ```
> 
> What is the purpose of `six.text_type` here @codecakes ?

Easy to google and understand

**Comment by milan-tom** [comment]
> > > @milan-tom do the following to fix the ☝️ issue and bring more structure and clarity to overall testing :
> > > 
> > > * create config/**init**.py folder and move all credentials you are creating in a `test_config.ini` file
> > > * Define sections as follows which will help you organize test data:
> > > 
> > > ```ini
> > > [DEFAULT]
> > > USERNAME = 
> > > PASSWORD = 
> > > EMAIL = 
> > > DB_USERNAME = 
> > > DB_PASSWORD = 
> > > 
> > > [verified]
> > > User =
> > > Password
> > > 
> > > [unverified]
> > > similarly here = 
> > > ```
> > > 
> > > 
> > > 
> > > * you can use `configparser` built-in lib to load from here like:
> > > 
> > > ```python
> > > import six
> > > import configparser
> > > 
> > > confg = configparser.ConfigParser()
> > > confg.read(your ini file)
> > > 
> > > user = six.text_type(confg['unverified']['USER']) and so on..
> > > ```
> > 
> > 
> > What is the purpose of `six.text_type` here @codecakes ?
> 
> Easy to google and understand

I did Google it. However, I just don't understand why we need it when the values are converted to strings by default. I have made a [solution](https://github.com/milan-tom/covidX/blob/add-test-cases/apps/auth_zero/tests.py) that does not require it.

**Comment by codecakes** [comment]
> > > > @milan-tom do the following to fix the ☝️ issue and bring more structure and clarity to overall testing :
> > > > 
> > > > * create config/**init**.py folder and move all credentials you are creating in a `test_config.ini` file
> > > > * Define sections as follows which will help you organize test data:
> > > > 
> > > > ```ini
> > > > [DEFAULT]
> > > > USERNAME = 
> > > > PASSWORD = 
> > > > EMAIL = 
> > > > DB_USERNAME = 
> > > > DB_PASSWORD = 
> > > > 
> > > > [verified]
> > > > User =
> > > > Password
> > > > 
> > > > [unverified]
> > > > similarly here = 
> > > > ```
> > > > 
> > > > 
> > > > 
> > > > * you can use `configparser` built-in lib to load from here like:
> > > > 
> > > > ```python
> > > > import six
> > > > import configparser
> > > > 
> > > > confg = configparser.ConfigParser()
> > > > confg.read(your ini file)
> > > > 
> > > > user = six.text_type(confg['unverified']['USER']) and so on..
> > > > ```
> > > 
> > > 
> > > What is the purpose of `six.text_type` here @codecakes ?
> > 
> > 
> > Easy to google and understand
> 
> I did Google it. However, I just don't understand why we need it when the values are converted to strings by default. I have made a [solution](https://github.com/milan-tom/covidX/blob/add-test-cases/apps/auth_zero/tests.py) that does not require it.


Because you are assuming a lot of things: the underlying code is guaranteed to be yours and it behaves a certain way. 
This will likely change tomorrow and we might not use .ini. That's why this [line in your code ](https://github.com/milan-tom/covidX/blob/d4e825c3c005dc55ac5f5d42f3b7ee7f1270dc5a/apps/auth_zero/tests.py#L11) looks convenient to you. 
What if the input was byte stream? 

Use six.ensure_text to ensure it gets the text input.  Break that line with correct interface definition like:

```python3

@dataclass(frozen=True)
class Credentials:
  user: str
  pass: str
  something_else: int

class CredentialsFactoryLoader:
  def __init__(self, config_file, credentials_class: credentials_class: TypeVar('Credentials', bound=True) = None, loader_class=None):
    self.credentials_class = credentials_class 
    self.loader = loader_class(config_file)

  def read_verified(self):
      return self.credentials_class(self.loader.read_verified_config())

  def read_unverified(self):
      return self.credentials_class(self.loader.read_unverified_config())  
 

class ConfigByteLoader():
  // some logic to load
  def sanitize(self):
    value = {}
     ensure whatever transformation you do here.
     ensure_text where text field expected
     ensure digits are converted
     return value dict here;
  def read_verified_config(self):
      pass

class ConfigFileLoader():
     def __init__(self, config_file: str):
       self.config_setting = config.read_file(config_file)
 
    def read_verified_config(self) -> dict:
      ensure whatever transformation you do here.
      ensure_text where text field expected
      ensure digits are converted
      return value dict here;
      return dict(user=self.config_settings['user'], pass=self.config_settings['pass'])
       

// older loader -> 
// creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigFileLoader) 
creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigByteLoader) 
create_user(**creds_obj.read_unverified())
```

File: apps/auth_zero/models.py [hunk 10]
```diff
@@ -26,10 +28,18 @@ class User(AbstractUser):
...
+    otp_verified = models.BooleanField(default=False)
+    email_verified = models.BooleanField(default=False)
 
     class Meta:
         app_label = "auth_zero"
 
+    @property
+    def is_verified(self):
+        return self.otp_verified and self.email_verified
+
```
Comment (codecakes) [inline apps/auth_zero/models.py:42]: @ElijahAhianyo no trailing spaces

File: apps/auth_zero/tests.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        self.assertTrue(fully_verified_user.otp_verified)
+        self.assertTrue(fully_verified_user.email_verified)
+        self.assertTrue(fully_verified_user.is_verified)
+
```
Comment (codecakes) [inline apps/auth_zero/tests.py:53]: @milan-tom run black. no trailing spaces

**Review by codecakes — CHANGES_REQUESTED** [review]
format files

File: apps/auth_zero/tests.py (hunk 11 above)
Comment (milan-tom) [reply apps/auth_zero/tests.py:53]: Sorry, this is the old code. The new code will be reflected once @ElijahAhianyo merges the pr in his repository.

**Comment by milan-tom** [comment]
> Because you are assuming a lot of things: the underlying code is guaranteed to be yours and it behaves a certain way.
> This will likely change tomorrow and we might not use .ini. That's why this [line in your code ](https://github.com/milan-tom/covidX/blob/d4e825c3c005dc55ac5f5d42f3b7ee7f1270dc5a/apps/auth_zero/tests.py#L11) looks convenient to you.
> What if the input was byte stream?
> 
> Use six.ensure_text to ensure it gets the text input. Break that line with correct interface definition like:
> 
> ```python
> @dataclass(frozen=True)
> class Credentials:
>   user: str
>   pass: str
>   something_else: int
> 
> class CredentialsFactoryLoader:
>   def __init__(self, config_file, credentials_class: credentials_class: TypeVar('Credentials', bound=True) = None, loader_class=None):
>     self.credentials_class = credentials_class 
>     self.loader = loader_class(config_file)
> 
>   def read_verified(self):
>       return self.credentials_class(self.loader.read_verified_config())
> 
>   def read_unverified(self):
>       return self.credentials_class(self.loader.read_unverified_config())  
>  
> 
> class ConfigByteLoader():
>   // some logic to load
>   def sanitize(self):
>     value = {}
>      ensure whatever transformation you do here.
>      ensure_text where text field expected
>      ensure digits are converted
>      return value dict here;
>   def read_verified_config(self):
>       pass
> 
> class ConfigFileLoader():
>      def __init__(self, config_file: str):
>        self.config_setting = config.read_file(config_file)
>  
>     def read_verified_config(self) -> dict:
>       ensure whatever transformation you do here.
>       ensure_text where text field expected
>       ensure digits are converted
>       return value dict here;
>       return dict(user=self.config_settings['user'], pass=self.config_settings['pass'])
>        
> 
> // older loader -> 
> // creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigFileLoader) 
> creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigByteLoader) 
> create_user(**creds_obj.read_unverified())
> ```

Could I use `django.utils.encoding.force_text` for this?

**Comment by codecakes** [comment]
> > Because you are assuming a lot of things: the underlying code is guaranteed to be yours and it behaves a certain way.
> > This will likely change tomorrow and we might not use .ini. That's why this [line in your code ](https://github.com/milan-tom/covidX/blob/d4e825c3c005dc55ac5f5d42f3b7ee7f1270dc5a/apps/auth_zero/tests.py#L11) looks convenient to you.
> > What if the input was byte stream?
> > Use six.ensure_text to ensure it gets the text input. Break that line with correct interface definition like:
> > ```python
> > @dataclass(frozen=True)
> > class Credentials:
> >   user: str
> >   pass: str
> >   something_else: int
> > 
> > class CredentialsFactoryLoader:
> >   def __init__(self, config_file, credentials_class: credentials_class: TypeVar('Credentials', bound=True) = None, loader_class=None):
> >     self.credentials_class = credentials_class 
> >     self.loader = loader_class(config_file)
> > 
> >   def read_verified(self):
> >       return self.credentials_class(self.loader.read_verified_config())
> > 
> >   def read_unverified(self):
> >       return self.credentials_class(self.loader.read_unverified_config())  
> >  
> > 
> > class ConfigByteLoader():
> >   // some logic to load
> >   def sanitize(self):
> >     value = {}
> >      ensure whatever transformation you do here.
> >      ensure_text where text field expected
> >      ensure digits are converted
> >      return value dict here;
> >   def read_verified_config(self):
> >       pass
> > 
> > class ConfigFileLoader():
> >      def __init__(self, config_file: str):
> >        self.config_setting = config.read_file(config_file)
> >  
> >     def read_verified_config(self) -> dict:
> >       ensure whatever transformation you do here.
> >       ensure_text where text field expected
> >       ensure digits are converted
> >       return value dict here;
> >       return dict(user=self.config_settings['user'], pass=self.config_settings['pass'])
> >        
> > 
> > // older loader -> 
> > // creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigFileLoader) 
> > creds_obj = CredentialsFactoryLoader(config_file, Credentials, ConfigByteLoader) 
> > create_user(**creds_obj.read_unverified())
> > ```
> 
> Could I use `django.utils.encoding.force_text` for this?

Yes, either one. six is compatibility module which we're already using. regardless, define interface so that we know what interface to conform to if we ever change the underlying abstraction. most likely, we will modify the config file soon;

File: common/config.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    def read_config(self, verification_status: str) -> dict:
+        return self.loader_class.read_config(verification_status)
+
+
+class ConfigByteLoader:
+    """Loads data from byte stream input"""
+
+    def __init__(self, config_file: str):
+        """Reads byte stream input as binary file."""
+        with open(config_file, "rb") as f:
```
Comment (codecakes) [inline common/config.py:37]: suggest to use IO input() -> StringIO() ->  open(io_file) as f and use this code snippet in a separate PR that focuses on testing and implementation of ConfigByteLoader

**Review by codecakes — CHANGES_REQUESTED** [review]

File: common/config.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+from dataclasses import dataclass
+from typing import TypeVar
+from django.utils.encoding import force_text
+
+
+@dataclass(frozen=True)
+class Credentials:
+    user: str
+    password: str
+    email: str
```
Comment (milan-tom) [inline common/config.py:11]: This class is not really used.

Comment (codecakes) [reply common/config.py:11]: so remove it @milan-tom

Comment (milan-tom) [reply common/config.py]: Done @codecakes

**Review by milan-tom — CHANGES_REQUESTED** [review]
All of @codecakes' requested changes have been completed. The only failing checks are due to the following reasons:

- Formatting problems in some migrations ➔ Based upon my understanding, migrations shouldn't be edited manually
- `1 blank line required before class docstring (found 0) (D203)` [error ](https://github.com/Xcov19/covidX/pull/69/files#annotation_868463850) from Codacy Static Code Analysis ➔ PEP 257 has removed this recommendation and so I think this check should be removed
- `Wildcard import base` ➔ This seems kind of unavoidable in this case as we want to build the dev settings file on top of the base settings file
-  `No name 'config' in module 'common'` ➔ Not sure why this error appears despite there being a `__init__.py` file in `common`
- `Unused variable 'acc_type_ids'` error in `apps/auth_zero/models.py` ➔ Perhaps we should remove this variable although it is not part of this pr

File: covidX/settings/dev.py [hunk 14]
```diff
@@ -0,0 +1,4 @@
+from .base import *
+
+DEBUG = True
+CONFIG_FILE = "apps/auth_zero/config/test_config.ini"
```
Comment (milan-tom) [inline covidX/settings/dev.py:4]: Should I move this config file out of the auth_zero directory and into the common directory?

File: apps/auth_zero/tests.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+            verification_status: self.user_model.objects.create_user(
+                **self.creds_obj.read_config(verification_status)
+            )
+            for verification_status in [
+                "fully_unverified",
+                "only_otp_verified",
+                "only_email_verified",
+                "fully_verified",
+            ]
+        }
```
Comment (milan-tom) [inline apps/auth_zero/tests.py:21]: Perhaps this should also be moved into the dev.py settings file so it does not have to be re-written for each test.

File: covidX/settings/dev.py (hunk 14 above)
Comment (codecakes) [reply covidX/settings/dev.py:4]: @milan-tom  no it should be part of configurations; may be the yaml config file in config/ and then for yaml file loader, it only loads specific to DEV/PREPROD/PROD settings

File: apps/auth_zero/tests.py (hunk 15 above)
Comment (codecakes) [reply apps/auth_zero/tests.py:21]: @milan-tom  configuring for test suites for several test cases is fine since setup is done only once for all test cases;

File: apps/auth_zero/tests.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
+from django.test import TestCase
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from common.config import ConfigFileLoader, CredentialsLoader
+
+
+class UserTestCase(TestCase):
+    def setUp(self):
+        self.user_model = get_user_model()
+        self.creds_obj = CredentialsLoader(settings.CONFIG_FILE, ConfigFileLoader)
```
Comment (codecakes) [inline apps/auth_zero/tests.py:10]: test_config config path can be moved and hard coded here. we will look at this later; but remove from dev.py and move it right here

**Review by codecakes — CHANGES_REQUESTED** [review]
look at branch `present-run ` and merge that here; then fix the suggested changes listed;
ill look at the semantic flow tomorrow

File: apps/auth_zero/tests.py (hunk 16 above)
Comment (milan-tom) [reply apps/auth_zero/tests.py]: Done

**Comment by marcelloromani** [comment]
@milan-tom > Formatting problems in some migrations ➔ Based upon my understanding, migrations shouldn't be edited manually

I have opened a bug about this. #84

**Comment by codecakes** [comment]
@ElijahAhianyo @milan-tom pls reformat with black:
apps/auth_zero/migrations/0004_auto_20201227_1514.py and;
apps/auth_zero/migrations/0005_user_email_verified.py

**Comment by codecakes** [comment]
reviewing this today

**Review by codecakes — CHANGES_REQUESTED** [review]
Until git diff conflicts are resolved I cannot review. This still needs to have correct changes from `present-run` branch;

File: common/config.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
+from configparser import ConfigParser
+from typing import TypeVar
+from django.utils.encoding import force_text
+
+
+class CredentialsLoader:
```
Comment (codecakes) [inline common/config.py:6]: Generally look at common/config.py here: https://github.com/Xcov19/covidX/pull/91/files
This is a better way to do what I've suggested below with interfaces;

File: apps/auth_zero/tests.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
+from django.test import TestCase
+from django.contrib.auth import get_user_model
+from common.config import ConfigFileLoader, CredentialsLoader
+
+
+class UserTestCase(TestCase):
+    def setUp(self):
+        self.user_model = get_user_model()
+        config_file = "apps/auth_zero/config/test_config.ini"
```
Comment (codecakes) [inline apps/auth_zero/tests.py:9]: Create data/config/test/test_config.ini instead of apps/auth_zero/config/test_config.ini

**Review by codecakes — CHANGES_REQUESTED** [review]
Left couple of comments for changes; I'll review rest post these changes;
Generally look at https://github.com/Xcov19/covidX/pull/91/files

1. Move common inside apps/
2. Create config/data/test/test_config.ini instead of apps/auth_zero/config/test_config.ini

File: apps/auth_zero/tests.py (hunk 18 above)
Comment (codecakes) [reply apps/auth_zero/tests.py:9]: what i mean by that is apps/auth_zero/config/data/test/test_config.ini instead of

**Comment by marcelloromani** [comment]
Tested the branch:

* created private key and certificate:
```
openssl req -x509 -newkey rsa:4096 -keyout privateKey.key -out certificate -days 365 -nodes
```

* created `.env` file based on values given by @codecakes privately

* started the application:
```
docker-compose up --build
```

* result: build errors:

```
(16:23:43) ERROR: /covidX/BUILD:16:11: Label '//covidX:settings/base.py' is invalid because 'covidX/settings' is a subpackage; perhaps you meant to put the colon here: '//covidX/settings:base.py'?
(16:23:43) ERROR: /BUILD:37:10: Target '//covidX:urls' contains an error and its package is in error and referenced by '//:manage'
(16:23:43) ERROR: /BUILD:37:10: Target '//covidX:asgi' contains an error and its package is in error and referenced by '//:manage'
(16:23:43) ERROR: /BUILD:37:10: Target '//covidX:wsgi' contains an error and its package is in error and referenced by '//:manage'
(16:23:43) ERROR: Analysis of target '//:manage' failed; build aborted: Analysis failed
```

**Review by codecakes — CHANGES_REQUESTED** [review]
Merge this feature/docker-custom-user-model
then request review;

**Comment by codecakes** [comment]
See https://github.com/Xcov19/covidX/pull/95 for reference what's changed

File: apps/auth_zero/tests.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
+from django.contrib.auth import get_user_model
+from django.test import TestCase
+
+from apps.common.config import ConfigFileLoader
+from apps.common.config import CredentialsLoader
```
Comment (milan-tom) [inline apps/auth_zero/tests.py:5]: @codecakes One of the conflicts was whether to use `from apps.common.config import` or `from common.config import`. Having previously removed the `__init__.py` file from `common`, as it is now within the `apps` package, I opted to use the former. Do you agree with this?

Comment (codecakes) [reply apps/auth_zero/tests.py:5]: __init__.py has nothing do with whether a package is standalone or not if you define a folder as a module; if a path is a part of import, then all modules in that path should have __init__.py

Comment (milan-tom) [reply apps/auth_zero/tests.py:5]: Okay. Thanks for the clarification. Am I right in thinking that I should still perform an absolute import even though I could import from `common`?

File: apps/auth_zero/tests.py [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class UserTestCase(TestCase):
+    def setUp(self):
+        self.user_model = get_user_model()
+        config_file = "apps/auth_zero/config/data/test/test_config.ini"
+        self.creds_obj = CredentialsLoader(config_file, ConfigFileLoader)
+        self.users = {
+            verification_status: self.user_model.objects.create_user(
```
Comment (codecakes) [inline apps/auth_zero/tests.py:14]: when you create a user/interact outside of application logic boundary, it is no longer a "unit" test; this is an integration test;
You could try using interfaces that mock or rather override functionality so that actual DB operations are mimicked and some stubs take over; we assume that they do certain things and we can define those implementations;

File: apps/auth_zero/tests.py (hunk 19 above)
Comment (codecakes) [reply apps/auth_zero/tests.py:5]: Always perform absolute imports

**Comment by marcelloromani** [comment]
https://github.com/PyCQA/pydocstyle/issues/141

Looks like in order to pass Codacy Static Code Analysis `D203` needs to be disabled.

File: apps/auth_zero/tests.py (hunk 20 above)
Comment (milan-tom) [reply apps/auth_zero/tests.py:14]: Could you give an example, please?

Comment (codecakes) [reply apps/auth_zero/tests.py:14]: Is this still an issue?

> 
> Could you give an example of how we could implement such interfaces in relation to the tests shown above, please?

**Comment by codecakes** [comment]
I have merged my branch; there are changes that need to be merged back here from master

**Comment by codecakes** [comment]
make sure `apps.common.config` absolute imports everywhere

File: apps/auth_zero/tests.py (hunk 20 above)
Comment (milan-tom) [reply apps/auth_zero/tests.py:14]: > Is this still an issue?
> 
> > Could you give an example of how we could implement such interfaces in relation to the tests shown above, please?

Do I need to rewrite the tests, or are we proceeding with the currently written tests?

**Review by milan-tom — COMMENTED** [review]
> make sure `apps.common.config` absolute imports everywhere

Are there any remaining `common.config` imports?

**Comment by codecakes** [comment]
> > make sure `apps.common.config` absolute imports everywhere
> 
> Are there any remaining `common.config` imports?

your code will throw errors on testing if there are any

File: apps/auth_zero/tests.py (hunk 20 above)
Comment (codecakes) [reply apps/auth_zero/tests.py:14]: > Do I need to rewrite the tests, or are we proceeding with the currently written tests?

If the program is not accessing anything outside its own boundary; if you can run a Django test in flight mode with no SQL/NoSQL connected, then its a unit test;

**Comment by codecakes** [comment]
> Hi @ElijahAhianyo I tested the branch:
> 
> * created private key and certificate:
> 
> ```
> openssl req -x509 -newkey rsa:4096 -keyout privateKey.key -out certificate -days 365 -nodes
> ```
> 
> * created `.env` file based on values given by @codecakes privately
> * started the application:
> 
> ```
> docker-compose up --build
> ```
> 
> * result: build errors:
> 
> ```
> (16:23:43) ERROR: /covidX/BUILD:16:11: Label '//covidX:settings/base.py' is invalid because 'covidX/settings' is a subpackage; perhaps you meant to put the colon here: '//covidX/settings:base.py'?
> (16:23:43) ERROR: /BUILD:37:10: Target '//covidX:urls' contains an error and its package is in error and referenced by '//:manage'
> (16:23:43) ERROR: /BUILD:37:10: Target '//covidX:asgi' contains an error and its package is in error and referenced by '//:manage'
> (16:23:43) ERROR: /BUILD:37:10: Target '//covidX:wsgi' contains an error and its package is in error and referenced by '//:manage'
> (16:23:43) ERROR: Analysis of target '//:manage' failed; build aborted: Analysis failed
> ```

@marcelloromani  this should be resolved now @milan-tom  ?

File: apps/auth_zero/tests.py (hunk 20 above)
Comment (milan-tom) [reply apps/auth_zero/tests.py:14]: Should I still work on this or will this not be applicable for the new repo?
